Provide ability to use AtomGroup as selection for Contact

shudipto-amin opened the pull request:
Fixes #2666

Summary
Solves the GSoC starter issue [#2666](https://github.com/MDAnalysis/mdanalysis/issues/2666)

Changes made in this Pull Request:
 - the class `mdanalysis/package/MDAnalysis/analysis/contacts.py::Contacts` is modified to accept static AtomGroups
 - the helper function `_new_selections()` in the same script is modified to use `Universe.copy()` **UPDATE: this has now been raised in new Issue #3570. This PR has reverted it back to the original.**
 - tests were added to `test_contacts.py` to ensure Contacts successfully uses static AtomGroups as well as selection texts.
 - tests were also added to ensure it fails when wrong types are provided for selection, including updating AtomGroups.

PR Checklist
------------
 - [x] Codecov coverage?
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

shudipto-amin committed:
Allow selections in Contacts to be AtomGroups

shudipto-amin committed:
Add tests for contacts implementation

shudipto-amin committed:
Fixes #2666

shudipto-amin committed:
Add author 

As part of solving the GSoC starter issue #2666

pep8speaks commented:
Hello @shudipto-amin! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:





There are currently no PEP 8 issues detected in this Pull Request. Cheers! :beers: 

##### Comment last updated at 2022-03-27 20:11:30 UTC

shudipto-amin committed:
Increase coverage of test.

hmacdope commented on testsuite/MDAnalysisTests/analysis/test_contacts.py:
```diff
@@ -171,16 +171,64 @@ class TestContacts(object):
+        assert_equal(with_atomgroup.grA, with_string.grA)
+        assert_equal(with_atomgroup.grB, with_string.grB) 
+
+    @pytest.mark.xfail
```
rather than using `xfail` it would be much better to use the `with pytest.raises()` construct with a match to the error message you are expecting (`select_error_message`).

shudipto-amin replied:
This is indeed a better approach! I updated the testing code to use `pytest.raises()`, and `pytest.mark.parametrize` to combine all expected failure tests in one function. As for referring to the `select_error_message`, I updated `contacts.Contacts` class to define it as a static attribute, and referred to it in `test_contacts.py` using `with pytest.raises(TypeError, match=contacts.Contacts.select_error_message)`. Is that okay, or is there a better way?

hmacdope commented on testsuite/MDAnalysisTests/analysis/test_contacts.py:
```diff
@@ -171,16 +171,64 @@ class TestContacts(object):
+        uag = universe.select_atoms(self.sel_acidic, updating=True)
+        self._run_Contacts(universe, sel_acidic=uag)
+    
+    @pytest.mark.xfail
```
same here with pytest.raises

hmacdope commented on testsuite/MDAnalysisTests/analysis/test_contacts.py:
```diff
@@ -171,16 +171,64 @@ class TestContacts(object):
+        uag = universe.select_atoms(self.sel_basic, updating=True)
+        self._run_Contacts(universe, sel_basic=uag)
+
+    @pytest.mark.xfail
```
same here with pytest.raises

hmacdope commented on testsuite/MDAnalysisTests/analysis/test_contacts.py:
```diff
@@ -171,16 +171,64 @@ class TestContacts(object):
         return mda.Universe(PSF, DCD)
 
-    def _run_Contacts(self, universe, start=None, stop=None, step=None, **kwargs):
+    def _run_Contacts(self, universe, sel_acidic=None, sel_basic=None, start=None, stop=None, step=None, **kwargs):
```
Can i just query why the choice of `sel_acidic` and `sel_basic`? perhaps I'm missing some context here.

shudipto-amin replied:
Thanks for all your comments! Let me answer this while I work on implementing pystest.raises(). 
I initially thought of using `ag_acidic` and `ag_basic`, where `ag` is for AtomGroup, but then I realized it doesn't have to be an AtomGroup, and can be a selection text. I thought `sel` would be more general. However, now I realize that it can be easily confused with `self.sel_basic` and `self.sel_acidic`. Should I change the variable names to `ag_*` instead of `sel_*`?

hmacdope replied:
yeah that makes sense.

hmacdope commented on package/CHANGELOG:
```diff
@@ -13,6 +13,13 @@ The rules for this file:
+
+Fixes Issue #2666
+* Contacts can now accept static AtomGroup as well as selection text.
+* helper function  `_new_selections()` in `contacts.py` no simply uses 
```
`no` vs `now` typo

hmacdope reviewed (changes requested):
This looks like real progress! See my comments for changes :)

IAlibay commented on package/CHANGELOG:
```diff
@@ -13,6 +13,13 @@ The rules for this file:
   * release numbers follow "Semantic Versioning" http://semver.org
 
 ------------------------------------------------------------------------------
+03/18/22 shudipto-amin
```
Unfortunately this is not how the changelog should be updated.

1. Update your branch so you have the latest version of the changelog
2. Add yourself at the current line 23 (on the line starting with my username)
3. Put fixes as a bullet point under Fixes (currently at line 28).

shudipto-amin replied:
Thanks for your instructions! I had some confusion about this part, which is clearer now. I updated the CHANGELOG accordingly, let me know if you have other corrections/suggestions.

shudipto-amin committed:
Resolve CHANGELOG conflicts and update it properly

tylerjereddy commented on testsuite/MDAnalysisTests/analysis/test_contacts.py:
```diff
@@ -171,16 +171,64 @@ class TestContacts(object):
+            except:
+                pass
+        if fails:
+            raise TypeError
```
Perhaps a bit awkward to catch all/generic `Exception`s in a test and manually raise a `TypeError` inside the test. Perhaps following the guidance related to `pytest.raises()` will clear that up to test for the expected errors in the code proper, since manually raising an error in the testsuite isn't of much use to end users of course.

If some of the parameters fail and others succeed you can xfail on a case-by-case basis as well.

shudipto-amin replied:
Thanks for you comment! I updated the code to use `@pytest.mark.parametrization` decorator, and I believe that should bring it in line with the style of `contacts.py`. I'm a bit uncomfortable with how I had to use the decorator, but maybe that's okay? Should we resolve?

shudipto-amin committed:
Tests expected failures with pytest.raises

* Replaces pytest.mark.xfail construct with pytest.raises construct.
* Uses pytest.mark.parametrize to test all expected fail cases in one function.

hmacdope commented on testsuite/MDAnalysisTests/analysis/test_contacts.py:
```diff
@@ -171,16 +171,61 @@ class TestContacts(object):
+        if ag2 == "USE UPDATING AG":
+            ag2 = universe.select_atoms(self.sel_basic, updating=True)
+
+        sel_err_msg = contacts.Contacts.select_error_message
```
No need to duplicate here, can just use `contacts.Contacts.select_error_message` directly in the `pytest.raises expression`

hmacdope commented:
@shudipto-amin looking good! Would you be able to resolve all the comments you have addressed so I can see what is still outstanding?

IAlibay commented on package/MDAnalysis/analysis/contacts.py:
```diff
@@ -387,14 +387,18 @@ class Contacts(AnalysisBase):
     """
+    # Error message for wrong selection type:
+    select_error_message = "selection must be either string or a " \
+                                    "static AtomGroup"
```
Please make sure you also review the other PEP8 issues that pep8speaks has highlighted: https://github.com/MDAnalysis/mdanalysis/pull/3565#issuecomment-1072770416

shudipto-amin committed:
Minor code clean up revisions

shudipto-amin committed:
Update CHANGELOG Fixes

IAlibay commented on package/CHANGELOG:
```diff
@@ -13,13 +13,19 @@ The rules for this file:
 Fixes
   * Updated the badge in README of MDAnalysisTest to point to Github CI Actions
-
+  * Provides Contacts ability to use static AtomGroups as well as selection 
```
Entries should be newest first in the changelog

shudipto-amin replied:
Done, I put it at the top.

IAlibay commented on package/MDAnalysis/analysis/contacts.py:
```diff
@@ -387,14 +387,18 @@ class Contacts(AnalysisBase):
        :class:`MDAnalysis.analysis.base.Results` instance.
     """
+    # Error message for wrong selection type:
+    select_error_message = "selection must be either string or a " \
+                           "static AtomGroup"
```
MDAnalysis tries to follow PEP8 where possible, as part of it, the recommendation is that line continuation should [use implicit line continuation where possible](https://peps.python.org/pep-0008/#:~:text=The%20preferred%20way%20of%20wrapping,a%20backslash%20for%20line%20continuation).

```suggestion
    select_error_message = ("selection must be either string or a "
                            "static AtomGroup")
```

shudipto-amin replied:
Fixed in commits made since this comment.

shudipto-amin committed:
Minor revision to CHANGELOG

IAlibay commented on package/MDAnalysis/analysis/contacts.py:
```diff
@@ -437,8 +441,26 @@ def __init__(self, u, select, refgroup, method="hard_cut", radius=4.5,
-        self.grA = u.select_atoms(select[0])
-        self.grB = u.select_atoms(select[1])
+
+        if isinstance(select[0], str):
+            self.grA = u.select_atoms(select[0])
+        elif isinstance(select[0], AtomGroup):
+            if isinstance(select[0], UpdatingAtomGroup):
+                raise TypeError(self.select_error_message)
+            else:
+                self.grA = select[0]
+        else:
+            raise TypeError(self.select_error_message)
+
+        if isinstance(select[1], str):
+            self.grB = u.select_atoms(select[1])
+        elif isinstance(select[1], AtomGroup):
+            if isinstance(select[1], UpdatingAtomGroup):
+                raise TypeError(self.select_error_message)
+            self.grB = select[1]
+        else:
+            raise TypeError(self.select_error_message)
```
There's a lot of repeated code here, let's see if we can condense it a little bit:

How about something like:

```python
@staticmethod
def _get_atomgroup(u, entry):
    if isinstance(entry, str):
        return u.select_atoms(entry)
    elif:
        if isinstance(entry, AtomGroup):
            if isinstance(entry, UpdatingAtomGroup):
                errmsg = "UpdatingAtomGroups are not supported"
                raise TypeError(errmsg)
            else:
                return entry
    else:
        raise TypeError(f"unrecognised selection type {entry}")

# ... and now back in __init__

self.grA, self.grB = [_get_atomgroup(var) for var in select]
```

This reduces the number of lines and also means you can directly test the staticmethod without having to go via class creation.

shudipto-amin replied:
Done! I added your suggested method in the Contacts class.

IAlibay commented on package/MDAnalysis/analysis/contacts.py:
```diff
@@ -492,7 +514,7 @@ def timeseries(self):
 def _new_selections(u_orig, selections, frame):
     """create stand alone AGs from selections at frame"""
-    u = MDAnalysis.Universe(u_orig.filename, u_orig.trajectory.filename)
+    u = u_orig.copy()
```
This technically changes the behaviour of q1q2, you could consider a case where `u` a trajectory which has been transferred in-memory and offset by a certain value (let's say you've shifted all the atomic positions by 10 A). By doing a copy on that object you now have a new trajectory that has the same shift in the atomic positions, whilst the old code would have led to a brand new universe that doesn't have this shift in atomic positions.

Whilst the copy behaviour is the one we are likely to want, introducing new changes that break behaviour without announcing it to users is a bad idea, because _someone_ might actually be relying on this behaviour (even though it's broken).

I realise @lilyminium pointed to fixing this in the original issue, but I don't agree that we should make this without warning users beforehand. I suspect further discussion will be needed here, so in the interest of saving time I would ask that we just not address this at this time and revisit it in the original issue before / fix it in a future PR.

shudipto-amin replied:
I see your point. From what you say, it does sound like it would be best to make a separate issue/PR out of it, and revert it back to original here. So should I go ahead and create a new issue for it and reference it here (as well as in #2666), or leave it to core developers to raise it?

IAlibay commented on testsuite/MDAnalysisTests/analysis/test_contacts.py:
```diff
@@ -171,16 +171,62 @@ class TestContacts(object):
         acidic = universe.select_atoms(self.sel_acidic)
         basic = universe.select_atoms(self.sel_basic)
+        if ag_acidic is None:
+            ag_acidic = self.sel_acidic
```
I see what you're doing here, but I think it actually makes things harder to read, I would suggestion creating a new fixture that's just for dealing with atomgroups.

shudipto-amin replied:
Done, all tests are done on new static method `Contacts._get_atomgroup()`

IAlibay commented on testsuite/MDAnalysisTests/analysis/test_contacts.py:
```diff
@@ -171,16 +171,62 @@ class TestContacts(object):
+            (sel_acidic, "USE UPDATING AG")
+        ]
+    )
+    def test_select_wrong_types(self, universe, ag1, ag2):
```
If going by my suggestion above, you can just directly do tests that run over the staticmethod instead here, it should be a lot simpler a parametrize.

IAlibay reviewed (changes requested):
Some comments, the way you've done it isn't wrong/bad, but it could be refined a little bit to make things easier long term.

shudipto-amin commented:
> @shudipto-amin looking good! Would you be able to resolve all the comments you have addressed so I can see what is still outstanding?

@hmacdope Thanks for your comments! I have resolved those conversations with the latest commits.

shudipto-amin committed:
Implements static method Contacts._get_atomgroup()

* Addresses IAlibay's comments:
    - Reduces code repetition
    - Makes test parameterization simpler by calling _get_atomgroup directly

shudipto-amin committed:
Correct CHANGELOG

shudipto-amin committed:
Create separate fixture for testing atomgroup

Previously, the `_run_Contacts()` of `test_contacts.py` was modified to allow
for atomgroup tests. Now, the `_run_Contacts()` is in it's original form, and
instead we use the newly created `_get_atomgroups()` function in `contacts.py`
to generate the atomgroups for testing purposes.

shudipto-amin committed:
PEP8 correction

shudipto-amin committed:
Change _new_selection function back to original

* As discussed in [this comment](https://github.com/MDAnalysis/mdanalysis/pull/3565#discussion_r831316421), the change to `_u_orig.copy()` with warning messages should be handled
in a different PR.

shudipto-amin commented:
@hmacdope @IAlibay @tylerjereddy 
Thank you all for your constructive comments! I learned a lot from them, and I believe I addressed all the concerns and the PR is ready or almost ready to merge. I just have a couple of questions:
1. According to [this discussion](https://github.com/MDAnalysis/mdanalysis/pull/3565#discussion_r831316421) with IAlibay, we should address the change to `universe.copy()` in a separate PR/Issue. Should I open a new issue for that?
2. The codecov report varies from time to time (without me pushing any new commits). Sometimes it shows the diff to be about -4%, sometimes 0%. Is there anything I can do to inverstigate/solve this, if it is a problem at all?

IAlibay commented:
> @hmacdope @IAlibay @tylerjereddy Thank you all for your constructive comments! I learned a lot from them, and I believe I addressed all the concerns and the PR is ready or almost ready to merge. I just have a couple of questions

Thanks, one of the mentors will re-review over the next few days.

> 1. According to [this discussion](https://github.com/MDAnalysis/mdanalysis/pull/3565#discussion_r831316421) with IAlibay, we should address the change to `universe.copy()` in a separate PR/Issue. Should I open a new issue for that?

Please do, refer the original issue please.

> 2. The codecov report varies from time to time (without me pushing any new commits). Sometimes it shows the diff to be about -4%, sometimes 0%. Is there anything I can do to inverstigate/solve this, if it is a problem at all?

This is likely due to us needing to approve the majority of the gh action runners when you do a new commit. It _shouldn't_ be an issue but we will look into it as part of the review process.

hmacdope commented on testsuite/MDAnalysisTests/analysis/test_contacts.py:
```diff
@@ -171,16 +171,43 @@ class TestContacts(object):
             **kwargs).run(start=start, stop=stop, step=step)
 
+    @pytest.mark.parametrize("seltxt", [sel_acidic, sel_basic])
+    def test_select_valid_types(self, universe, seltxt):
```
Could you add one more test, almost the same as this one but where you actually `.run()` the contacts with a `string` and `AtomGroup` input and compare the results?


You have tested the internal `staticmethod` `Contacts.get_atomgroup()` but not the actual flow of data through the whole function when a `string ` vs `AtomGroup` is used.

shudipto-amin replied:
Done! See [line 199:](https://github.com/MDAnalysis/mdanalysis/pull/3565/files#diff-c79b8e25558e1ba2f042d25eaa0e355502afba57cf03bf8c9b22e0c6103063a4R199)
```python
def test_contacts_selections(self, universe):
```

hmacdope reviewed (changes requested):
Almost there @shudipto-amin just one more test IMO then should be good to go. Thanks for your hard work.

hmacdope commented:
Don't worry about codecov, it is just the runners not auto-triggering for first time contributors.

shudipto-amin committed:
Add test function for Contacts workflow

IAlibay commented on package/MDAnalysis/analysis/contacts.py:
```diff
@@ -387,14 +387,31 @@ class Contacts(AnalysisBase):
+                             "are not supported.")
+
+    @staticmethod
+    def _get_atomgroup(u, sel):
```
If possible could this be moved after `__init__`?

Usually folks inspecting classes will want to see the initialisation as the first thing so that they can quickly get an idea of what the class expects.

IAlibay commented on package/MDAnalysis/analysis/contacts.py:
```diff
@@ -387,14 +387,31 @@ class Contacts(AnalysisBase):
        :attr:`timeseries` results are now stored in a
        :class:`MDAnalysis.analysis.base.Results` instance.
     """
```
Please do add a `versionchanged` entry that mentions that AtomGroups are now supported.

shudipto-amin replied:
Done

IAlibay commented on package/MDAnalysis/analysis/contacts.py:
```diff
@@ -387,14 +387,31 @@ class Contacts(AnalysisBase):
        :class:`MDAnalysis.analysis.base.Results` instance.
     """
+    # Error message for wrong selection type:
+    _select_error_message = ("selection must be either string or a "
```
This is now only needed within `_get_atomgroup` so I would just define it there rather than making it a class attribute.

shudipto-amin replied:
I actually do use it in [this line](https://github.com/MDAnalysis/mdanalysis/pull/3565/files#diff-c79b8e25558e1ba2f042d25eaa0e355502afba57cf03bf8c9b22e0c6103063a4R228) of `test_contacts.py`:
```python
       with pytest.raises(
            TypeError, match=contacts.Contacts._select_error_message
        ) as te:
            contacts.Contacts._get_atomgroup(universe, ag)
```
I can't think of another way to access that error message that doesn't require first calling a class/method. Is there some way, or do I just keep it as in internal attribute?

IAlibay replied:
Honestly I think it's better to:
a) not have this as a class attribute
b) make the error messages within `_get_atomgroup` a bit more specific if you feel like it
c) just write a subsection of the message as part of the match in `test_select_wrong_types`

Convenience variables are great, but having them stick around for the single test just isn't worth it.

shudipto-amin replied:
Okay, done!

IAlibay commented on testsuite/MDAnalysisTests/analysis/test_contacts.py:
```diff
@@ -171,16 +171,65 @@ class TestContacts(object):
+    ):
         acidic = universe.select_atoms(self.sel_acidic)
         basic = universe.select_atoms(self.sel_basic)
+
```
Improving readability is appreciated but I also don't think it's worth affecting the git history for an empty line.
```suggestion
```

IAlibay reviewed (changes requested):
Couple more things

shudipto-amin committed:
Revert to previous git history

Address [this comment](https://github.com/MDAnalysis/mdanalysis/pull/3565/files#r833332107)

Co-authored-by: Irfan Alibay <IAlibay@users.noreply.github.com>

shudipto-amin committed:
Move _get_atomgroup() to after __init__ in Contacts

shudipto-amin committed:
Update versionchanged in contacts.py

shudipto-amin reviewed (commented):
@IAlibay I have addressed all but one comment below, regarding `_select_error_message` attribute, which I believe is still needed for testing functions.

@hmacdope I added the workflow test as you requested, [see line 199](https://github.com/MDAnalysis/mdanalysis/pull/3565/files#diff-c79b8e25558e1ba2f042d25eaa0e355502afba57cf03bf8c9b22e0c6103063a4R199):
```python
    def test_contacts_selections(self, universe):
```

IAlibay commented on testsuite/MDAnalysisTests/analysis/test_contacts.py:
```diff
@@ -181,6 +184,51 @@ def _run_Contacts(self, universe, start=None, stop=None, step=None, **kwargs):
+            universe,
+            select=(aga, agb),
+            refgroup=(aga, agb),
+        ).run()
```
It's very pedantic but I'm going to object to these a little bit, this is not very readable.

If you're going to call run on it in one line then please re-arrange so that it's easier to read.

Otherwise close the brackets at 208 and call `cag.run()` separately.

Same thing below please.

shudipto-amin replied:
Opted for rearrangement. Is this better?

IAlibay replied:
Sorry my readability issue wasn't with the way the arguments were arranged but instead with the `).run()` being alone on a single line. The compactness advantage of calling `run()` at the same time as class construction just disappears in this case.

shudipto-amin replied:
I think I get you now. I put `.run()` on a new line altogether in [this latest commit](https://github.com/MDAnalysis/mdanalysis/pull/3565/commits/55636bf304a590b0c6e52dfee76a13e0d4861a48#diff-c79b8e25558e1ba2f042d25eaa0e355502afba57cf03bf8c9b22e0c6103063a4R215).

shudipto-amin committed:
Make code more readable

Addresses these comments by IAlibay:
- [Comment on readability](https://github.com/MDAnalysis/mdanalysis/pull/3565/files#r833749271)
- [Comment on convenience variable](https://github.com/MDAnalysis/mdanalysis/pull/3565/files#r833748384)

shudipto-amin committed:
Make code more readable

* put .run() on a separate line

IAlibay commented on package/MDAnalysis/analysis/contacts.py:
```diff
@@ -386,15 +386,18 @@ class Contacts(AnalysisBase):
        :attr:`timeseries` results are now stored in a
        :class:`MDAnalysis.analysis.base.Results` instance.
+    .. versionchanged:: 2.2.0
+       :class: `Contacts` accepts both AtomGroup and string for `select`
```
Unfortunately read the docs hasn't triggered but there shouldn't be a blank space between :class: and the classname.
```suggestion
       :class:`Contacts` accepts both AtomGroup and string for `select`
```

IAlibay commented on package/CHANGELOG:
```diff
@@ -13,14 +13,16 @@ The rules for this file:
   * Fixed DumpReader triclinic box dimensions reading. (Issue #3386, PR #3403)
   * Updated the badge in README of MDAnalysisTest to point to Github CI Actions
-
+  
```
Can you keep this empty line but remove the whitespace from it?

shudipto-amin replied:
```suggestion
 
```

IAlibay commented on testsuite/MDAnalysisTests/analysis/test_contacts.py:
```diff
@@ -181,6 +184,51 @@ def _run_Contacts(self, universe, start=None, stop=None, step=None, **kwargs):
+        ag_from_string = contacts.Contacts._get_atomgroup(universe, seltxt)
+        ag_from_ag = contacts.Contacts._get_atomgroup(universe, ag)
+
+        assert_equal(ag_from_string, ag_from_ag)
```
atomgroup comparison already calls np.array_equal in the background, doing a standard assert is orders of magnitude faster here (seconds to microseconds).
```suggestion
        assert ag_from_string == ag_from_ag
```

IAlibay reviewed (changes requested):
Just a few tiny things and then I'm good with this.

@hmacdope do you want to re-review please?

shudipto-amin reviewed (commented):
Minor edits via suggestions.

shudipto-amin committed:
Apply suggestions from code review

Changes:
* Comparing atom groups directly with `assert` instead of `assert_equal`, which is faster.
* Minor formatting edits.

Co-authored-by: Irfan Alibay <IAlibay@users.noreply.github.com>

shudipto-amin committed:
Minor change to CHANGELOG

hmacdope reviewed (approved):
Looks good to me @shudipto-amin. Thanks for all your hard work, we appreciate your contribution.

shudipto-amin commented:
@IAlibay and @hmacdope, once again thanks for the feedback. Is there anything else, or is this ready to merge?

IAlibay commented:
I think it's waiting on a review from me, will do in a little bit

IAlibay commented:
Cycling PR to restart gh action runners that had halted last Friday (gh infrastructure issues it appears)

IAlibay reviewed (approved):
ltgm, thanks for the work on this @shudipto-amin.
